src/Featurizers/LagLeadOperatorFeaturizer.h

```diff
@@ -0,0 +1,412 @@
+            *std::max_element(_offsets.cbegin(), _offsets.cend())))
+    ) {
+        // prepopulate circular buffer with null values to imitate non-existing prior rows
+        for (std::int64_t i = 0; i < (static_cast<std::int64_t>(_horizon) - _offset_delta - 1); ++i) {
+            _buffer.push(Traits<TransformedTargetType>::CreateNullValue());
+        }
+}
+
+template <typename T>
+LagLeadOperatorTransformer<T>::LagLeadOperatorTransformer(Archive &ar) :
+    LagLeadOperatorTransformer(
+        [&ar](void) {
+            // Version
+            std::uint16_t                   majorVersion(Traits<std::uint16_t>::deserialize(ar));
+            std::uint16_t                   minorVersion(Traits<std::uint16_t>::deserialize(ar));
+
+            if(majorVersion != 1 || minorVersion != 0)
+                throw std::runtime_error("Unsupported archive version");
+
+            // Data
+            std::uint32_t horizon = Traits<std::uint32_t>::deserialize(ar);
+            std::vector<std::int64_t> offsets = Traits<std::vector<std::int64_t>>::deserialize(ar);
+            
+            return LagLeadOperatorTransformer(std::move(horizon), std::move(offsets));
+        }()
+    ) {
+}
+
+template <typename T>
+bool LagLeadOperatorTransformer<T>::operator==(LagLeadOperatorTransformer const &other) const {
+    return this->_horizon == other._horizon && this->_offset_delta == other._offset_delta
+            && this->_offsets == other._offsets;
+}
+
+template <typename T>
+bool LagLeadOperatorTransformer<T>::operator!=(LagLeadOperatorTransformer const &other) const {
+    return (*this == other) == false;
+}
+
+template <typename T>
+void LagLeadOperatorTransformer<T>::save(Archive &ar) const /*override*/ {
+    // Version
+    Traits<std::uint16_t>::serialize(ar, 1); // Major
+    Traits<std::uint16_t>::serialize(ar, 0); // Minor
+
+    // Data
+    Traits<std::uint32_t>::serialize(ar, _horizon);
+
+    // reset _offsets to its input version
+    std::vector<std::int64_t> offsets(_offsets);
+    for (auto it = offsets.begin(); it != offsets.end(); it++) {
+        *it += _offset_delta;
+    }
+    Traits<std::vector<std::int64_t>>::serialize(ar, offsets);
+
+    // Note that we aren't serializing working state
+}
+
+template <typename T>
+void LagLeadOperatorTransformer<T>::execute_helper(TransformedTargetType const &input, typename BaseType::CallbackFunction const &callback) {
+    _buffer.push(input);
+    // until the circular buffer is full, we don't have enough data to generate output matrix
+    if (!_buffer.is_full()) {
+        return;
+    }
+    typename BaseType::TransformedType ret(_offsets.size(), _horizon);
+    
+    for (std::uint32_t row = 0; row < _offsets.size(); ++row) {
+        std::tuple<typename Components::CircularBuffer<TransformedTargetType>::iterator, typename Components::CircularBuffer<TransformedTargetType>::iterator> range{_buffer.range(_horizon, static_cast<size_t>(_offsets[row]))};
```

Comment:
>static_cast<size_t>(_offsets[row]) [](start = 189, length = 34)

This cast is working around a potential issue. Conceptually speaking, an offset should never be > 0. A delta can be less than 0. So, incoming deltas are converted to offsets in the constructor. Deltas should be a signed type, offsets should not unsigned.

Reply:
Good point! The transformer is now keeping offsets as uint32_t now.